log/logtest/recorder.go

```diff
@@ -55,87 +74,102 @@ func NewRecorder(options ...Option) *Recorder {
-
-	// Records are the log records, and their associated context this
-	// instrumentation scope recorded.
-	Records []EmittedRecord
 }
 
-// EmittedRecord holds a log record the instrumentation received, alongside its
-// context.
-type EmittedRecord struct {
-	log.Record
+// Record represents the record alongside its context.
+type Record struct {
+	// Ensure forward compatibility by explicitly making this not comparable.
+	_ [0]func()
 
-	ctx context.Context
+	Context           context.Context
+	EventName         string
+	Timestamp         time.Time
+	ObservedTimestamp time.Time
+	Severity          log.Severity
+	SeverityText      string
+	Body              log.Value
+	Attributes        []log.KeyValue
 }
 
-// Context provides the context emitted with the record.
-func (rwc EmittedRecord) Context() context.Context {
-	return rwc.ctx
-}
-
-// Recorder is a recorder that stores all received log records
-// in-memory.
-type Recorder struct {
-	embedded.LoggerProvider
-
-	mu      sync.Mutex
-	loggers []*logger
-
-	// enabledFn decides whether the recorder should enable logging of a record or not
-	enabledFn enabledFn
+// Clone returns a deep copy.
+func (a Record) Clone() Record {
+	b := a
+	attrs := make([]log.KeyValue, len(a.Attributes))
+	copy(attrs, a.Attributes)
+	b.Attributes = attrs
+	return b
 }
 
 // Logger returns a copy of Recorder as a [log.Logger] with the provided scope
 // information.
 func (r *Recorder) Logger(name string, opts ...log.LoggerOption) log.Logger {
 	cfg := log.NewLoggerConfig(opts...)
-
-	nl := &logger{
-		scopeRecord: &ScopeRecords{
-			Name:       name,
-			Version:    cfg.InstrumentationVersion(),
-			SchemaURL:  cfg.SchemaURL(),
-			Attributes: cfg.InstrumentationAttributes(),
-		},
-		enabledFn: r.enabledFn,
+	scope := Scope{
+		Name:       name,
+		Version:    cfg.InstrumentationVersion(),
+		SchemaURL:  cfg.SchemaURL(),
+		Attributes: cfg.InstrumentationAttributes(),
 	}
-	r.addChildLogger(nl)
 
-	return nl
-}
-
-func (r *Recorder) addChildLogger(nl *logger) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 
-	r.loggers = append(r.loggers, nl)
+	if r.loggers == nil {
+		r.loggers = make(map[Scope]*logger)
+	}
+
+	l, ok := r.loggers[scope]
+	if ok {
+		return l
+	}
+	l = &logger{
+		enabledFn: r.enabledFn,
+	}
+	r.loggers[scope] = l
+	return l
 }
 
-// Result returns the current in-memory recorder log records.
-func (r *Recorder) Result() []*ScopeRecords {
+// Result returns a deep copy of the current in-memory recorded log records.
+func (r *Recorder) Result() Recording {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 
-	ret := []*ScopeRecords{}
-	for _, l := range r.loggers {
-		ret = append(ret, l.scopeRecord)
+	res := make(Recording, len(r.loggers))
+	for s, l := range r.loggers {
+		func() {
+			l.mu.Lock()
+			defer l.mu.Unlock()
+			if l.records == nil {
+				res[s] = nil
+				return
+			}
+			recs := make([]Record, len(l.records))
+			for i, r := range l.records {
+				recs[i] = r.Clone()
+			}
+			res[s] = recs
+		}()
```

Comment:
just wondering, this is wrapped with an anonymous func so we can use `return` instead of `continue`? or is there another reason for this?

Reply:
Another. Defer unlock of the mutex